backend/pixeldrain/api_client.go

```diff
@@ -0,0 +1,374 @@
+		return fs.ErrorPermissionDenied
+	}
+
+	return e
+}
+
+func paramsFromMetadata(meta fs.Metadata) (params url.Values) {
+	params = make(url.Values)
+
+	if modified, ok := meta["mtime"]; ok {
+		params.Set("modified", modified)
+	}
+	if created, ok := meta["btime"]; ok {
+		params.Set("created", created)
+	}
+	if mode, ok := meta["mode"]; ok {
+		params.Set("mode", mode)
+	}
+	if shared, ok := meta["shared"]; ok {
+		params.Set("shared", shared)
+	}
+	if loggingEnabled, ok := meta["logging_enabled"]; ok {
+		params.Set("logging_enabled", loggingEnabled)
+	}
+
+	return params
+}
+
+// nodeToObject converts a single FilesystemNode API response to an object. The
+// node is usually a single element from a directory listing
+func (f *Fs) nodeToObject(node FilesystemNode) (o *Object) {
+	// Trim the path prefix. The path prefix is hidden from rclone during all
+	// operations. Saving it here would confuse rclone a lot. So instead we
+	// strip it here and add it back for every API request we need to perform
+	node.Path = strings.TrimPrefix(node.Path, f.pathPrefix)
+	return &Object{fs: f, base: node}
+}
+
+func (f *Fs) nodeToDirectory(node FilesystemNode) fs.DirEntry {
+	return fs.NewDir(strings.TrimPrefix(node.Path, f.pathPrefix), node.Modified)
+}
+
+func (f *Fs) escapePath(p string) (out string) {
+	// Add the path prefix, encode all the parts and combine them together
+	var parts = strings.Split(f.pathPrefix+p, "/")
+	for i := range parts {
+		parts[i] = url.PathEscape(parts[i])
+	}
+	return strings.Join(parts, "/")
+}
+
+func (f *Fs) put(
+	ctx context.Context,
+	path string,
+	body io.Reader,
+	meta fs.Metadata,
+	options []fs.OpenOption,
+) (node FilesystemNode, err error) {
+	var params = paramsFromMetadata(meta)
+
+	// Tell the server to automatically create parent directories if they don't
+	// exist yet
+	params.Set("make_parents", "true")
+
+	resp, err := f.srv.CallJSON(
+		ctx,
+		&rest.Opts{
+			Method:     "PUT",
+			Path:       f.escapePath(path),
+			Body:       body,
+			Parameters: params,
+			Options:    options,
+		},
+		nil,
+		&node,
+	)
+	if err != nil {
+		return node, err
+	}
+	return node, resp.Body.Close()
+}
+
+func (f *Fs) read(ctx context.Context, path string, options []fs.OpenOption) (in io.ReadCloser, err error) {
+	resp, err := f.srv.Call(ctx, &rest.Opts{
+		Method:  "GET",
+		Path:    f.escapePath(path),
+		Options: options,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, err
+}
+
+func (f *Fs) stat(ctx context.Context, path string) (fsp FilesystemPath, err error) {
+	resp, err := f.srv.CallJSON(
+		ctx,
+		&rest.Opts{
+			Method: "GET",
+			Path:   f.escapePath(path),
+			// To receive node info from the pixeldrain API you need to add the
+			// ?stat query. Without it pixeldrain will return the file contents
+			// in the URL points to a file
+			Parameters: url.Values{"stat": []string{""}},
+		},
+		nil,
+		&fsp,
+	)
+	if err != nil {
+		return fsp, err
+	}
+	return fsp, resp.Body.Close()
```

Comment:
It is probably preferable to put `NoResponse: true` in the `rest.Opts` rather than explicitly close the body then it definitely gets closed and it will be consistent with the other calls.

Reply:
Wouldn't the response get discarded if I set NoReponse here?

Reply:
Looks like I mis-read the code!

`CallJSON` with a non-nil response shouldn't need to have the body closed so I think you can remove the `resp.Body.Close()`.

https://github.com/rclone/rclone/blob/e334366345440af8c8392d47e662a8bc3607a4a4/lib/rest/rest.go#L449-L450